polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -7,43 +7,71 @@ use polars_time::prelude::*;
 use super::*;
 
 #[cfg(feature = "date_offset")]
-pub(super) fn date_offset(s: Series, offset: Duration) -> PolarsResult<Series> {
+pub(super) fn date_offset(s: &[Series]) -> PolarsResult<Series> {
+    let sa = &s[0];
+    let sb = &s[1];
+
     let preserve_sortedness: bool;
-    let out = match s.dtype().clone() {
+    let out = match sa.dtype().clone() {
         DataType::Date => {
-            let s = s
+            let sa = sa
                 .cast(&DataType::Datetime(TimeUnit::Milliseconds, None))
                 .unwrap();
             preserve_sortedness = true;
-            date_offset(s, offset).and_then(|s| s.cast(&DataType::Date))
+            date_offset(&[sa, sb.clone()]).and_then(|s| s.cast(&DataType::Date))
```

Comment:
instead of cloning and recursing, would it work to just work with `.date()`?

Something like:
```Rust
   let ts = &s[0];
    let offsets = &s[1].utf8().unwrap();

    let preserve_sortedness: bool;
    let out = match ts.dtype().clone() {
        DataType::Date => {
            const MSECS_IN_DAY: i64 = MILLISECONDS * SECONDS_IN_DAY;
            let date = ts.date().unwrap();
            let out = match offsets.len() {
                1 => {
                    let offset = match offsets.get(0) {
                        Some(offset) => Duration::parse(offset),
                        _ => Duration::new(0),
                    };
                    date.0.try_apply(|v| {
                        Ok(
                            (Duration::add_ms(&offset, MSECS_IN_DAY * v as i64, None)? / MSECS_IN_DAY)
                                as i32,
                        )
                    })
                }
                _ => unreachable!(),
            }?;
            preserve_sortedness = offsets.len() == 1;
            out.cast(&DataType::Date)
        }
        DataType::Datetime(tu, tz) => {
            let datetime = ts.datetime().unwrap();
```

Reply:
I'm not entirely sure why the recursive approach is used - I just tried to modify the current code as little as possible.